QuestionAnswering/QA_Agent/deployment/fastapi_app/services/nlq2action/execute_action/sparql/process_response.py

```diff
@@ -0,0 +1,101 @@
+from functools import cache
+from typing import Annotated, Dict, List, Union
+
+from fastapi import Depends
+
+from services.entity_store import EntityStore, get_entity_store
+from services.support_data import DataItem, TableDataItem, WktDataItem
+
+
+class SparqlResponseProcessor:
+    TWA_PREFIXES = [
+        "http://www.theworldavatar.com/kg/",
+        "https://www.theworldavatar.com/kg/",
+    ]
+    WKT_LITERAL_PREFIX = "<http://www.opengis.net/def/crs/OGC/1.3/CRS84> "
```

Comment:
Linked to the other comment, but be aware that this is not the only CRS that WKT literals can come in, although I would generally encourage using this one when possible as it is the default and is effectively the same as EPSG:4326.

Reply:
Is there a specification for WKT literals that I can refer to? So that I can safely parse WKT literals into approprate data objects. Or is there a library that I can use for this?

Reply:
Unfortunately the GeoSPARQL standard doesn't actually support CRSs and officially assumes that all WKT literals are in CRS84. Ontop extends the official standard by allowing an optional CRS IRI to be appended to the start of the WKT literal. An example of some of these can be seen in [this OBDA file](https://github.com/ontop/ontop/blob/170318de14b1affdeeab43bf3fe2dee5cd0cf9e7/binding/owlapi/src/test/resources/geosparql/geosparql-h2.obda#L10).

I'd probably recommend forcing developers to reproject their non-EPSG:4326 geometries to EPSG:4326 in the `source` query of their OBDA mappings using the PostGIS [`ST_Transform`](https://postgis.net/docs/ST_Transform.html) function and not specifying the CRS IRIs in the WKTLiterals. This will save confusion as Ontop is not able to perform operations on geometries in different projections and you won't need to parse out the CRS IRIs.

Reply:
Is the projection to EPSG:4326 and removal of CRS IRI from WKTLiteral a to-do item that you would suggest to be done asap, following which Zaha should not check for CRS IRI and accept the WKTLiteral verbatim? Then I'll need to coordinate with Kok Foong to make the change. 

Otherwise, the current code works for both scenarios when `<http://www.opengis.net/def/crs/OGC/1.3/CRS84>` is prepended and when it is not, so no changes are needed (unless the code must also anticipate other CRS IRIs).

Reply:
Keeping the code as it is is fine for now, I just wanted to make sure you were aware of the potential issues.

Reply:
@kok-foong pointed me to Jena library for parsing WKT literal, specifically the method [`WKTReader::extract`](https://github.com/apache/jena/blob/cf3a14008b7cbcf5abc8cf485eec4501566f8469/jena-geosparql/src/main/java/org/apache/jena/geosparql/implementation/parsers/wkt/WKTReader.java#L255). Referencing the Jena implementation, I wrote a simple [method to parse WKT literals](https://github.com/cambridge-cares/TheWorldAvatar/blob/fa0166847fd769940388bec56559619ca7d4b824/QuestionAnswering/QA_Agent/deployment/fastapi_app/services/wkt.py).

If the SRS URI is found to be CRS84, then the WKT will be visualised. Otherwise, it will be presented verbatim in a table. The code for this handling is [here](https://github.com/cambridge-cares/TheWorldAvatar/blob/fa0166847fd769940388bec56559619ca7d4b824/QuestionAnswering/QA_Agent/deployment/fastapi_app/controllers/qa/execute_action/sparql/process_response.py#L83).